text/0000-otlp-trace-data-format.md

```diff
@@ -0,0 +1,361 @@
+
+**Status:** `proposed`
+
+OTLP Trace Data Format specification describes the structure of the trace data that is transported by OpenTelemetry Protocol (RFC0035).
+
+## Motivation
+
+This document is a continuation of OpenTelemetry Protocol RFC0035 and is necessary part of OTLP specification.
+
+## Explanation
+
+OTLP Trace Data Format is primarily inherited from OpenCensus protocol. Several changes are introduced with the goal of more efficient serialization. Notable differences from OpenCensus protocol are:
+
+1. Removed `Node` as a concept.
+2. Extended `Resource` to better describe the source of the telemetry data.
+3. Replaced attribute maps by lists of key/value pairs.
+4. Eliminated unnecessary additional nesting in various values.
+
+Changes 1-2 are conceptual, changes 3-4 improve performance.
+
+## Internal details
+
+This section specifies data format in Protocol Buffers.
+
+### Resource
+
+```
+// Resource information. This describes the source of telemetry data.
+message Resource {
+  // Set of labels that describe the resource. See OpenTelemetry specification
+  // semantic conventions for standardized label names:
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md
+
+  repeated AttributeKeyValue labels = 3;
+  int32 dropped_labels_count = 11;
+}
+```
+
+### Span
+
+```
+// A span represents a single operation within a trace. Spans can be
+// nested to form a trace tree. Spans may also be linked to other spans
+// from the same or different trace. And form graphs. Often, a trace
+// contains a root span that describes the end-to-end latency, and one
+// or more subspans for its sub-operations. A trace can also contain
+// multiple root spans, or none at all. Spans do not need to be
+// contiguous - there may be gaps or overlaps between spans in a trace.
+//
+// The next id is 17.
+message Span {
+  // A unique identifier for a trace. All spans from the same trace share
+  // the same `trace_id`. The ID is a 16-byte array. An ID with all zeroes
+  // is considered invalid.
+  //
+  // This field is semantically required. Receiver should generate new
+  // random trace_id if empty or invalid trace_id was received.
+  //
+  // This field is required.
+  bytes trace_id = 1;
+
+  // A unique identifier for a span within a trace, assigned when the span
+  // is created. The ID is an 8-byte array. An ID with all zeroes is considered
+  // invalid.
+  //
+  // This field is semantically required. Receiver should generate new
+  // random span_id if empty or invalid span_id was received.
+  //
+  // This field is required.
+  bytes span_id = 2;
+
+  // This field conveys information about request position in multiple distributed tracing graphs.
+  // It is a list of Tracestate.Entry with a maximum of 32 members in the list.
+  //
+  // See the https://github.com/w3c/distributed-tracing for more details about this field.
+  message Tracestate {
```

Comment:
* what is the value of explicitly parsing this instead of using the full `tracestate` header as a single string?
* what is the value of storing this in the span at all?

Reply:
This is a good point. The structure is inherited from OC proto and I can see that for example OC Collector / OTel Collector decode/encode/translate this field from/to HTTP header however unclear _why_ they do it.
It seems unnecessary work since the field is never examined or modified (which would justify having it in structured form).
@bogdandrutu perhaps you have more info about why having this in structured form is useful?

Reply:
> since the field is never examined or modified

I guess custom propagators might have logic that involves the actual trace state values.

Reply:
@Oberon00 can you please expand on that? I do not quite understand what a "custom propagator" is.

Reply:
Maybe I should have written "formatters" instead of "propagators", as I guess that's currently the customization point. E.g. imagine instead of the W3C TraceContext I need to use a custom propagation format. That format might require a "level" HTTP header in addition to the parent ID and trace flags. That level field would be a integer that is zero for a root span 1 for its remote children, 2 for their remote children and so on. To implement that, I would store that level in the TraceState and increment it each time the state is propagated using a custom HTTPTextFormat implementation.

Reply:
What is more, at least in Python the TraceState is represented as the equivalent of a `HashMap<String, String>` anyway.

Reply:
My recommendation is to drop this field from this PR. I don't see a strong motivation/use case for keeping it in the out-of-band data format. Can always add it later.

Reply:
@yurishkuro I am not sure I understand what "out-of-band" means in this context, please clarify.

I think the field itself is needed to be able to properly translate (without data loss) and forward spans received from HTTP-based formats (or any other formats that use w3c tracestate). If this field does not exist in this protocol how would such translation work? We would have nowhere to place the tracestate and would have to drop it.

I am thinking about Collector use case which may receive for example Zipkin HTTP with w3c tracestate in the header and forward in OTLP.

I may be missing something, please correct me.

--
I agree that it is most likely unnecessary to keep tracestate in a structured form here. It may be good enough to simply have an opaque byte array or a string (not even a map - as @Oberon00 mentioned is done in Python). So far I see no arguments in favor of any structured form. Any parties that are interested in the internal structure of tracestate are free to decode, modify and encode it according to w3c recommendation.

Reply:
In-band is the data that's propagated with the request in the headers, such as W3C trace context. That's where tracestate exists.

Out-of-hand is the performance data (spans) reported asynchronously, off the critical path of the request, to the tracing backend. Is there really a need for that data to include `tracestate`? (maybe there is, but it should be documented why).

Reply:
> Out-of-hand is the performance data (spans) reported asynchronously, off the critical path of the request, to the tracing backend. Is there really a need for that data to include `tracestate`? (maybe there is, but it should be documented why).

This is a good question, which I also would like to understand and which is unclear to me by reading w3c definition of `tracestate`. I raised an issue asking for a clarification and listed 2 use cases of how I think a tracing backend could use the `tracestate`: https://github.com/w3c/trace-context/issues/346

I would like to know though if my understanding is correct (and whether it is inline with w3c understanding).

Reply:
To make sure we are not stuck on this single item I suggest that we merge the PR and I will address this in a separate PR.
I am resolving this comment, if you disagree please reopen.